src/bodies/rapier_area.rs

```diff
@@ -201,217 +311,171 @@ impl RapierArea {
-            self.state.monitored_objects.insert(
-                handle_pair_hash,
-                MonitorInfo {
-                    id: other_area_id,
-                    instance_id: other_area_instance_id,
-                    object_shape_index: other_area_shape as u32,
-                    area_shape_index: area_shape as u32,
-                    collision_object_type: CollisionObjectType::Area,
-                    state: 1,
-                },
-            );
-            space
-                .get_mut_state()
-                .area_add_to_monitor_query_list(self.base.get_id());
-        }
+        self.state
+            .unhandled_event_queue
+            .insert(other_collider_handle, new_event);
+        // Finally, we tell the space that we have events in our queue.
+        space
+            .get_mut_state()
+            .area_add_to_monitor_query_list(self.base.get_id());
     }
 
-    #[allow(clippy::too_many_arguments)]
-    pub fn on_area_exit(
+    // Returns true if the new event is required (eg it doesn't contradict or invalidate an existing unhandled event).
+    fn filter_on_unhandled_events(
         &mut self,
-        collider_handle: ColliderHandle,
-        other_area: &mut Option<&mut RapierCollisionObject>,
-        other_area_shape: usize,
-        other_area_id: RapierId,
-        other_area_instance_id: u64,
-        area_collider_handle: ColliderHandle,
-        area_shape: usize,
         space: &mut RapierSpace,
-    ) {
-        if self.area_monitor_callback.is_none() {
-            return;
-        }
-        if let Some(other_area) = other_area
-            && let Some(other_area) = other_area.get_mut_area()
-            && !other_area.is_monitorable()
+        other_collider_handle: ColliderHandle,
+        new_event_state: i32,
+    ) -> bool {
+        if let Some(current_event_in_queue) =
+            self.state.unhandled_event_queue.get(&other_collider_handle)
         {
-            return;
-        }
-        // Remove from currently detected areas
-        if let Some(detected_area) = self.state.detected_areas.get_mut(&other_area_id) {
-            *detected_area -= 1;
-            if *detected_area == 0 {
-                self.state.detected_areas.remove(&other_area_id);
-            }
-        } else {
-            return;
-        }
-        let handle_pair_hash = (collider_handle, area_collider_handle);
-        if let Some(monitored_object) = self.state.monitored_objects.get(&handle_pair_hash) {
-            if monitored_object.state != 1 {
-                godot_error!("Area is not being monitored");
+            // See if we currently have a pending unhandled event (eg an event from this frame) that is the opposite of this one.
+            // If so, this event will cancel that one out.
+            if current_event_in_queue.state == -new_event_state {
+                self.state
+                    .unhandled_event_queue
+                    .remove(&other_collider_handle);
```

Comment:
Is this a queue or a set? If it's a queue, will this remove all ocurrences of it from queue?